### ISSUE TITLE
Describe the Weblate REST API with OpenAPI

### ISSUE THREAD

**Issue by walpox** [WeblateOrg/weblate#12584]
### Describe the problem

There is not a problem to solve per se. The [current API documentation is good](https://docs.weblate.org/en/latest/api.html). In my opinion, the proposed OpenAPI standard might be an improvement upon such documentation.

### Describe the solution you would like

[OpenAPI](https://swagger.io/specification/) (formerly known as Swagger) is one of the most popular specifications to describe REST APIs formally. It allows developers to define the entire Weblate REST API with a root OpenAPI document.

Some advantages this standard might offer over the current API documentation are:

- Make it easier for developers to code their own third-party REST clients to connect to the Weblate REST API, by using tooling which generates code automatically based on the OpenAPI document.
- Make it easier to document the Weblate REST API:
  - The OpenAPI specification provides a standard way to document REST APIs. It should be easier to document the API when you can use a well-known syntax and keywords (structure), without having the API documentation written in free-form (`.rst` files).
  - Sphinx seems to have plugins to integrate with OpenApi. If they support version 3.1 of the specification, it should be possible to display the OpenAPI document directly within the Weblate documentation.
- It should facilitate automating testing of the Weblate REST API.

### Screenshots

This is how the Weblate REST API looks in Swagger UI:

![demo-weblate-api-swagger-ui](https://github.com/user-attachments/assets/377d5f96-9ce3-4269-9a59-e80e5e058d89)

### Additional context

I already wrote the first draft of a OpenAPI document for the Weblate REST API, as a learning experience of the spec. I'm linking a PR here later.

**Comment by nijel** [comment]
https://github.com/WeblateOrg/weblate/pull/12612 integrates drf-spectacular to Weblate. This gives some OpenAPI schema, but it is very likely incomplete and needs additional annotations to make it really a replacement for the current documentation. But I think it's the correct direction that will allow us to discard manually maintained documentation in the future.

**Comment by nijel** [comment]
I've made some cleanups in https://github.com/WeblateOrg/weblate/pull/12645 and https://github.com/WeblateOrg/weblate/pull/12644, but there is still a lot to review and fix.

**Comment by nijel** [comment]
I've also added CI action to lint - https://github.com/WeblateOrg/weblate/actions/workflows/api.yml. It shows quite some errors and warnings for now: https://github.com/WeblateOrg/weblate/actions/runs/11126455162

**Comment by walpox** [comment]
It looks promising. I'm going to keep proposing changes to make the documentation better.

**Comment by nijel** [comment]
https://github.com/WeblateOrg/weblate/pull/14246 is the next step. It uses OpenAPI in the documentation instead of manually maintaining it. It will need additional steps:

* Review the documentation if something is missing from the current docs.
* Update links from the rest of the documentation to the OpenAPI one (it uses different references for URLs).
* Remove the old docs.

I'm not sure what would be the best approach now. Do we want to merge #14246 and iterate over the remaining steps, or wait until we have all the bits ready?

**Comment by walpox** [comment]
Presently, are we talking about changes to the documentation contained in [docs.weblate.org](https://docs.weblate.org/en/latest/index.html), the one built with Read the Docs and Sphinx?
I am asking to verify we are not talking about changes to other places such as the API documentation provided by a deployed Weblate instance, like [hosted.weblate.org/api/docs/](https://hosted.weblate.org/api/docs/).

If so, I would like to ask what the desired end goal is here. For example, let's say, to replace the current [REST API page](https://docs.weblate.org/en/latest/api.html) with content generated automatically from the OpenAPI schema (with `sphinxcontrib-openapi`). To improve that page by getting rid of manual maintenance and relying solely on Weblate's API codebase for this documentation.

After this is discussed and the picture a bit clearer, I will look into the steps mentioned in your comments, for example, for missing information in the documentation generated by the OpenAPI schema.

**Comment by nijel** [comment]
Yes, the goal is to avoid manually documenting the API. The pull request adds it to the API page duplicating the content temporarily.

**Comment by nijel** [comment]
Here it is rendered: https://weblate--14246.org.readthedocs.build/en/14246/api.html#openapi

**Comment by walpox** [comment]
Sorry for the late response, I have been busy with some other stuff...

I noticed some information is missing when using the Sphinx plugin for OpenAPI, such as the string in `info.description`.

I will review the codebase of this plugin to see if this is intended or rather a bug, and make appropriate fixes.

Thus, this PR should be moved to a later milestone because I don't think I will have time to review and prepare the changes for the next release of Weblate.

**Comment by nijel** [comment]
I've already removed it from the milestone as I've noticed the missing bits as well. Some of them are definitely caused by https://github.com/sphinx-contrib/openapi/issues/107.

**Comment by nijel** [comment]
The https://github.com/sphinx-contrib/openapi repo is not really receiving much of a maintenance, so I'm not sure if this is a viable approach. Therefore, I've modified https://github.com/WeblateOrg/weblate/pull/14246 to only keep track of the OpenAPI spec in the repository, but not include it in the documentation.

We still need to figure out how to deal with that. I'd honestly love to have the documentation based on the OpenAPI specs. I see some approaches:

* Use redoc only for the API documentation. This would make it hard to link from the rest of the documentation to the API one.
* Bring https://github.com/sphinx-contrib/openapi back to life by making it support OpenAPI 3 features.
* Add own conversion from OpenAPI to Sphinx to cover the bits we need.

**Comment by walpox** [comment]
I will have some time this weekend to spend on this issue and I will then provide you some feedback. Sorry for the delay, not having much time recently...

**Comment by walpox** [comment]
As you say, it seems the `sphinxcontrib-openapi` extension doesn't have much active development.

In my opinion, having our own extension or fork to implement the missing functionality and render the contents inside Sphinx seems to be the way to go.

I will evaluate me starting development of this extension or fork by the end of the year. In the meantime, I suggest keeping the documentation as it is. At least, the OpenAPI spec is still available at `/api/docs`, with a good enough format.

**Comment by walpox** [comment]
I have finished the other project. Thus, I have free time I will use to work on this issue from now on.

If you want to add anything while I work on this (e.g. changed requirements), feel free to do so.

**Comment by nijel** [comment]
Thanks! The only new think that popped up is https://github.com/WeblateOrg/weblate/pull/15729. As mentioned there I don't think this is a good approach, but some of the points are definitely valid.

**Comment by walpox** [comment]
I would like to let you know I am working on this extension from time to time, it's not abandoned.

I should be able to get a production-ready release later this year. I'm doing it from scratch as a learning exercise and it's going well.

<img alt="Image" src="https://github.com/user-attachments/assets/7632c81b-8f01-44fc-b479-eb71f6d6bdf7" />

**Comment by nijel** [comment]
That looks great!

**Comment by nijel** [comment]
Short update from my side. In process of validating wlc API usage, several cleanups to the OpenAPI schema are in progress:

* https://github.com/WeblateOrg/weblate/pull/19130
* https://github.com/WeblateOrg/weblate/pull/19144

**Comment by nijel** [comment]
@walpox Any update on this? Perhaps we can help with this? The manually maintained API documentation is annoying, as it is super easy to miss updating it (like seen in https://github.com/orgs/WeblateOrg/discussions/19867).

**Comment by walpox** [comment]
It's still a work in progress. Hopefully, a release stable enough for production will later.

I will keep working on the project now that I have more spare time. I uploaded the repo here:

<https://codeberg.org/walpo/sphinx-openapi-renderer>

It is a prerelease, so it should not be used in production for now. It has errors, bugs, etc.

**Comment by nijel** [comment]
Thanks for sharing that. Meanwhile, https://github.com/sphinx-contrib/openapi/issues/180 shows some interest in the original extension as well; perhaps it would be good to cooperate on a shared codebase.
